librad/src/id/entity.rs

```diff
@@ -0,0 +1,515 @@
+    }
+    pub fn to_json_writer<W>(&self, writer: W) -> Result<(), Error>
+    where
+        W: std::io::Write,
+    {
+        self.to_data().to_json_writer(writer)?;
+        Ok(())
+    }
+    pub fn to_json_string(&self) -> Result<String, Error> {
+        self.to_data().to_json_string()
+    }
+
+    pub fn from_json_reader<R>(r: R) -> Result<Self, Error>
+    where
+        R: std::io::Read,
+    {
+        Self::from_data(data::EntityData::from_json_reader(r)?)
+    }
+    pub fn from_json_str(s: &str) -> Result<Self, Error> {
+        Self::from_data(data::EntityData::from_json_str(s)?)
+    }
+
+    pub fn compute_hash(&self) -> Result<Multihash, Error> {
+        Ok(Sha2_256::digest(&self.canonical_data()?))
+    }
+
+    pub async fn check_user_key(
+        &self,
+        user_uri: &RadicleUri,
+        key: &PublicKey,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        if !self.has_key(key) {
+            return Err(Error::KeyNotPresent(key.to_owned()));
+        }
+        let user = resolver.resolve(user_uri).await?;
+        if !user.has_key(key) {
+            return Err(Error::UserKeyNotPresentWIP(
+                user_uri.to_owned(),
+                key.to_owned(),
+            ));
+        }
+        Ok(())
+    }
+
+    pub async fn check_key(
+        &self,
+        key: &PublicKey,
+        by: &Signatory,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        match by {
+            Signatory::OwnedKey => {
+                if !self.has_key(key) {
+                    return Err(Error::KeyNotPresent(key.to_owned()));
+                }
+            },
+            Signatory::User(uri) => {
+                let user = resolver.resolve(&uri).await?;
+                if !user.has_key(key) {
+                    return Err(Error::UserKeyNotPresentWIP(uri.to_owned(), key.to_owned()));
+                }
+            },
+        }
+        Ok(())
+    }
+
+    pub fn compute_signature(&self, key: &Key) -> Result<Signature, Error> {
+        Ok(key.sign(&self.canonical_data()?))
+    }
+
+    pub async fn sign(
+        &mut self,
+        key: &Key,
+        by: &Signatory,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        let public_key = key.public();
+        if self.signatures().contains_key(&public_key) {
+            return Err(Error::SignatureAlreadyPresent(public_key.to_owned()));
+        }
+        self.check_key(&public_key, by, resolver).await?;
+        let signature = EntitySignature {
+            by: by.to_owned(),
+            sig: self.compute_signature(key)?,
+        };
+        self.signatures.insert(public_key, signature);
+        Ok(())
+    }
+
+    pub async fn check_signature(
+        &self,
+        key: &PublicKey,
+        by: &Signatory,
+        signature: &Signature,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        self.check_key(key, by, resolver).await?;
```

Comment:
iiuc, this checks that the _most recent_ user identity has this public key (ie. it was not revoked). If I'm not mistaken, `check_history` will fail then if any signature valid in the past was revoked.